### DISCUSSION TITLE
on_each_cpu function can't be rightly included

### DISCUSSION THREAD

**Discussion by wanggeyang** [XKCP/XKCP#85] [Q&A] [answered]
I'm recently working on profiling and optimizing the performace of android ARM binaries. Now I want to know the cpu cycles consumed by each instruction, but I encountered a problem when using the on_each_cpu function. The header file that defines the on_each_cpu function is linux/smp.h, but I found that this header file seems to have been deprecated in the latest version of NDK. When I downloaded the lower version of NDK(r15c) and included the linux/smp.h, I got the following error:

              fatal error: 'linux/smp.h' file not found
              #include <linux/smp.h>
                       ^~~~~~~~~~~~~

then, I added the LOCAL_C_INCLUDES field pointing to the folder where smp.h is located to the Android.mk file,  the following errors appeared which seems to be incompatible.

	     warning: '__ANDROID_API__' macro redefined [-Wmacro-redefined]
              #define __ANDROID_API__ 19

              <built-in>:359:9: note: previous definition is here
              #define __ANDROID_API__ 21

              error: no member named 'vswscanf' in the global namespace
              using ::vswscanf;

              error: unknown type name 'u16'
              typedef struct { volatile u16 offset[256]; } __regbase16;
                                        ^
              
              fatal error: 'omap730.h' file not found
              #include "omap730.h"
                       ^~~~~~~~~~~

The added LOCAL_C_INCLUDES is: 
LOCAL_C_INCLUDES := $(LOCAL_PATH) \
$(LOCAL_PATH)/../../../android-ndk-r15c/platforms/android-19/arch-arm/usr/include

So i want to know which version of NDK are you using, and why your smp.h and on_each_cpu function can be rightly included and rightly run.

**Comment by BrunoPrho** [comment, marked as answer]
Hi,

The Kernel PMU never been tried on Android OS and not with NDK. So, I would expect regarding the specificity some issues. 
- The module is at kernel level and so fully linked to the OS. The dependency on the linux headers is quasi a constraint even for some unix distro not delivering natively all components and sources. The module is then loaded with `sudo` permission.
- I can see that the Android 12 Google "native" still use the `"smp.h"` and `on_each_cpu`[1].  But I am not sure all flavored Android do so.
- I had a quick look to see the Android methodology for `.ko` compilation and it sounds from far pretty much similar to the `makefile` . See some link [2,3,4,5]. However, I notice that many do cross compilation while the Kernel PMU is compiled on the target. It's sounds easier on SOC platform.
- I have installed AndoidStudio and if you can let me know more detail (target device, the env you are using, how you compile, how you get access to an android terminal) I might give it a try to the condition you are planning to use the XKCP. 

On another note:
- the XKCP has a new bench marking performance based on time convert back to cycle which will give you the performance of any algorithm in the package. 
 -  If you want to bench specific ARM instructions, you will need another tool to run a sufficient number an ARM ASM instruction to get an estimate.

Hope it helps, below some references mentioned.


[1] https://android.googlesource.com/kernel/common/+/refs/tags/android12-5.10-2021-07_r1/include/linux/smp.h
[2]https://source.android.com/devices/architecture/kernel/loadable-kernel-modules
[3] https://github.com/julius-b/android-lkm
[4]https://sites.google.com/a/mtu.edu/max-jin-home-page/hints/buildandloadakernelmoduleinandroidsystem/build-and-load-a-kernel-module-in-android-system
[5] https://tthtlc.wordpress.com/2011/12/29/how-to-write-a-kernel-module-on-android-sony-ericsson-xperia-arc-s/
